### PR TITLE
Obtaining client cert for registered user by CSR

### PR DESCRIPTION
**Pull request by lucafavatellawork**
This is the postman API doc PR #154 rebased on master, plus the related code. The test is in https://github.com/opensparkl/sse_cli/pull/82 (the sse_cli repo contains "black"-box-style tests for this sse_core repo, docs, and a Python test library): see `claim_cert_and_privkey.rst` (regression test) and `claim_cert_by_csr.rst` (for this new functionality).

TODO:
- [x] Check merged postman files can be opened.
- [x] Add deprecation notice in postman for cert requests that make private key generated by node.
- [x] Comments in code.
- [x] Comment that cert call only starts generation - it does not synchronously generate cert.
- [ ] ~Check if tempfile handling needs refactoring.~
  - No refactoring seems needed on this regard.
- [x] Comment assumption on tempfile name without whitespaces.
- [x] Review usage of `?RETURN` macro.
  - Opened #160
- [x] Consider validating user in CSR.
- [x] Consider validating node in CSR.
- [x] Consider validating attributes in CSR.

Not TODO:
- Race conditions in client certificate issuance.
- Client certificate renewal.

### CODE REVIEW

**Comment by lucafavatellawork** [comment]
Up to here this is the postman API doc PR #154 rebased on master.

**Comment by lucafavatellawork** [comment]
https://github.com/opensparkl/sse_core/pull/159/commits/d94fa88078da26725827929676a24a2baac13f15 in this PR was tested successfully by [CI on sse_cli repo](https://travis-ci.com/github/opensparkl/sse_cli/builds/168450403#L249).

File: apps/sse_yaws/src/sse_yaws_tls.erl [hunk 1]
```diff
@@ -272,6 +274,144 @@ get_client_certs(UserName, CertFormat) ->
...
+    %%   {2,5,4,3},
+    %%   <<12,15,111,119,110,101,114,64,108,111,99,97,108,104,111,115,116>>}],
+    %% [{'AttributeTypeAndValue',
+    %%   {2,5,4,10},
+    %%   <<12,23,115,115,101,49,64,52,52,53,99,49,98,49,52,49,49,53,53,46,108,111,99,97,108>>}]]
+    %% ```
+    {ok, ActualRawUserName} =
+      get_attribute_value(
+        ?'id-at-commonName', ReqSubject),
+    <<12, 15, ActualUserName/binary>> =
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_tls.erl:347]: This is wrong as it hardcodes length 15 (from DER).

File: apps/sse_yaws/src/sse_yaws_tls.erl [hunk 2]
```diff
@@ -272,6 +274,144 @@ get_client_certs(UserName, CertFormat) ->
...
+      ActualRawUserName,
+    {true, _} =
+      {string:equal(UserName, ActualUserName),
+        {?'id-at-commonName', {UserName, ActualUserName}}},
+    {ok, ActualRawNodeName} =
+      get_attribute_value(
+        ?'id-at-organizationName', ReqSubject),
+    <<12, 23, ActualNodeName/binary>> =
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_tls.erl:355]: This is wrong as it hardcodes length 23 (from DER).

File: apps/sse_yaws/src/sse_yaws_tls.erl [hunk 3]
```diff
@@ -272,6 +274,167 @@ get_client_certs(UserName, CertFormat) ->
...
+  {[], _} =
+    {As2, subject},
+
+  {ok, CommonName, OrganizationName}.
+
+%% @doc
+%% Starts the generation of a client certificate
+%% for the supplied user name.
+%% If the cert file for that user has been generated already,
+%% return error.
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_tls.erl:329]: Enforce at most one cert generation as revocation list is not hooked in TLS server (yet).

**Comment by lucafavatellawork** [comment]
https://github.com/opensparkl/sse_core/pull/159/commits/f3f398400004b1b9100167ad3aa476bc79bb896d on this PR passed [CI running the corresponding test](https://travis-ci.com/github/opensparkl/sse_cli/builds/168631006#L249).

File: apps/sse_yaws/src/sse_yaws_tls.erl (hunk 3 above)
Comment (jacobythwaites) [reply apps/sse_yaws/src/sse_yaws_tls.erl:329]: Yes, this is out of scope

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 4]
```diff
@@ -335,6 +359,115 @@ client_cert(Arg, _User) ->
...
+
+  {ok, TempReqFile} =
+    sse_util:tempfile(),
+  try
+    {ok, ReqFd} =
+      file:open(TempReqFile, [write]),
+    try
+      ok =
+        file:write(ReqFd, ClientReq),
+      ok
```
Comment (jacobythwaites) [inline apps/sse_cfg/src/sse_cfg_yapp.erl:417]: Unnecessary line, but reads fine.

File: apps/sse_yaws/priv/openssl/gen_client_certs.sh [hunk 5]
```diff
@@ -21,7 +21,7 @@
 # on the local filesystem only.
 #
 # Usage:
-#   sh gen_client_certs.sh <work_dir> <username> <node_name>
+#   sh gen_client_certs.sh <work_dir> <username> <node_name> <req_file>
```
Comment (jacobythwaites) [inline apps/sse_yaws/priv/openssl/gen_client_certs.sh:24]: Nice enhancement

**Pull request merged**